server/src/main/java/org/elasticsearch/index/engine/IdStoredFieldLoader.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.index.engine;
+
+import org.apache.lucene.index.CodecReader;
+import org.apache.lucene.index.FieldInfo;
+import org.apache.lucene.index.LeafReader;
+import org.apache.lucene.index.StoredFieldVisitor;
+import org.elasticsearch.common.CheckedBiConsumer;
+import org.elasticsearch.index.mapper.IdFieldMapper;
+import org.elasticsearch.index.mapper.Uid;
+
+import java.io.IOException;
+
+/**
+ * Loads IDs from stored fields using a stored fields reader optimized for sequential access
+ */
+final class IdStoredFieldLoader {
+
+    private final CheckedBiConsumer<Integer, StoredFieldVisitor, IOException> reader;
+    private final IdOnlyFieldVisitor visitor = new IdOnlyFieldVisitor();
+
+    /**
+     * Creates a new ID reader over a segment
+     */
+    IdStoredFieldLoader(LeafReader in) {
+        this.reader = getStoredFieldsReader(in);
+    }
+
+    private static CheckedBiConsumer<Integer, StoredFieldVisitor, IOException> getStoredFieldsReader(LeafReader in) {
+        if (in instanceof CodecReader) {
```

Comment:
We wrap the directory reader in # so this is never true.  We don't use codec readers, instead we introduced a `SequentialStoredFieldsLeafReader` that exposes the merge instance lightly (see [`SourceLookup`]()).
That avoids rewriting all filter leaf readers as codec readers. And so in this case the [LeafReaderWithLiveDocs] should extend `SequentialStoredFieldsLeafReader` to expose this functionality.

Reply:
Good catch! Thanks Jim!